### DISCUSSION TITLE
Why is the performance of extensions written with zephir not as good as native php?

### DISCUSSION THREAD

**Discussion by 010test** [zephir-lang/zephir#2367] [Q&A] [answered]
hello erveryone, I am a chinese beginner in zephir, and I want to write an example to actually test the performance of zephir. The result puzzles me. My code is as follows.

php code:

```
function getRandStr($length) {
    $str = 'abcdefghijklmnopqrstuvwxyz123456789';
    return substr(str_shuffle($str), 0, $length);
}
```

zephir code:

```
namespace HelloWord;
 
class Greeting
{
    public static function say(int len)
    {
            var str = "abcdefghijklmnopqrstuvwxyz123456789";
            return substr(str_shuffle(str), 0, len);
    }
}
```

test code:

```
$st = de();
for ($i = 0; $i <= 10000000; $i++) {
    HelloWord\Greeting::say(6) . "\n";
}
$en = de();

echo 'zephir: ', $en - $st;
echo "\r\n";

$st = de();
for ($i = 0; $i <= 10000000; $i++) {
    getRandStr(6);
}
$en = de();

echo 'php: ', $en - $st;
echo "\r\n";

function de() {
    [$usec, $sec] = explode(" ", microtime());
    return ((float)$usec + (float)$sec);

}
```

The result is that zephir basically takes more than 3 seconds to run, while native php only takes more than 2 seconds. This is the result of multiple tests, even if it is run separately or the order is changed.

Please tell me where I wrote this wrong or is zephir just a performance problem? thanks a lot !

**Comment by Jeckerson** [comment, marked as answer]
Hello, because string concatenation won't have any performance gain. Your code will use native PHP functions, but with +1 Zephir wrapper. In that cases you will need to implement your own functions written in C and called by zephir.

See: https://docs.zephir-lang.com/0.12/en/optimizers

Also see: https://github.com/zephir-lang/zephir/issues/2151

**Reply by 010test** [reply]
ok, thanks a lot, i will write more test to test zephir performance
